### PR TITLE
Add ability to parse from ships.json, a hierarchical unit structure

### PR DESCRIPTION
**Pull request by royfalk**
This is a first attempt to reduce units.json file size by moving ships to another file and making it hierarchical.
The aim is to write a identical value once. For example, all variants of the same ship (e.g. llama) share the folder, name, description, etc. If you write these in the parent llama, all variants will apply them. Have a different description for llama.begin, overwrite it for that variant specifically.

Currently, implemented only for Areus (and only for Areus variant) and Llama.begin.

See also related asset [98](https://github.com/vegastrike/Assets-Production/pull/98)
Code Changes:
- [not yet] Have the PR Validation Tests been run? See https://github.com/vegastrike/Vega-Strike-Engine-Source/wiki/Pull-Request-Validation

Issues:
- ~~In loaded game, shield reduce in power~~ (fixed)
- In loaded game, Oswald started as neutral or hostile (non reproducible)
- When re-spawned after dying, I was in an unidentified ship with no mesh and icepicks for weapons. Couldn't dock or die either. (non reproducible)

### CODE REVIEW

**Comment by stephengtuggy** [comment]
With regard to the issues you noted: I've seen problems with shields not regenerating to full capacity with current `master`, as well. I think this was introduced with one of the recent PRs.

**Comment by stephengtuggy** [comment]
Looks like the Windows builds are failing; we'll need to take a look at that.

**Comment by stephengtuggy** [comment]
The build errors on Windows are as follows:

```
  Error: D:\a\Vega-Strike-Engine-Source\Vega-Strike-Engine-Source\engine\src\cmd\unit_optimize_factory.h(34): error C2039: 'map': is not a member of 'std'
  C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.33.31629\include\fstream(26): note: see declaration of 'std'
  Error: D:\a\Vega-Strike-Engine-Source\Vega-Strike-Engine-Source\engine\src\cmd\unit_optimize_factory.h(34): error C2061: syntax error: identifier 'map'
  Error: D:\a\Vega-Strike-Engine-Source\Vega-Strike-Engine-Source\engine\src\cmd\unit_optimize_factory.cpp(30): error C2511: 'void UnitOptimizeFactory::RecuriveParse(std::map<std::string,std::string,std::less<std::string>,std::allocator<std::pair<const std::string,std::string>>>,const std::string &,bool)': overloaded member function not found in 'UnitOptimizeFactory'
  D:\a\Vega-Strike-Engine-Source\Vega-Strike-Engine-Source\engine\src\cmd\unit_optimize_factory.h(30): note: see declaration of 'UnitOptimizeFactory'
  Error: D:\a\Vega-Strike-Engine-Source\Vega-Strike-Engine-Source\engine\src\cmd\unit_optimize_factory.cpp(57): error C2660: 'UnitOptimizeFactory::RecuriveParse': function does not take 3 arguments
  D:\a\Vega-Strike-Engine-Source\Vega-Strike-Engine-Source\engine\src\cmd\unit_optimize_factory.h(34): note: see declaration of 'UnitOptimizeFactory::RecuriveParse'
  Error: D:\a\Vega-Strike-Engine-Source\Vega-Strike-Engine-Source\engine\src\cmd\unit_optimize_factory.cpp(82): error C2660: 'UnitOptimizeFactory::RecuriveParse': function does not take 3 arguments
  D:\a\Vega-Strike-Engine-Source\Vega-Strike-Engine-Source\engine\src\cmd\unit_optimize_factory.h(34): note: see declaration of 'UnitOptimizeFactory::RecuriveParse'
```

File: engine/src/cmd/unit_optimize_factory.h [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Vega Strike.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include "vsfilesystem.h"
+
```
Comment (BenjamenMeyer) [inline engine/src/cmd/unit_optimize_factory.h:29]: missing the `#include <map>` - which would probably resolve the various errors

**Review by BenjamenMeyer — CHANGES_REQUESTED** [review]
Missing a header include;  always good practice to have any dependent headers directly included.

**Comment by royfalk** [comment]
> With regard to the issues you noted: I've seen problems with shields not regenerating to full capacity with current master, as well. I think this was introduced with one of the recent PRs.

Opened #743 to track this.

File: engine/src/cmd/unit_optimize_factory.h (hunk 1 above)
Comment (royfalk) [reply engine/src/cmd/unit_optimize_factory.h:31]: Done and fixed.

File: engine/src/cmd/unit_optimize_factory.h [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+#include <map>
+
+#include "vsfilesystem.h"
+
+class UnitOptimizeFactory
+{
+public:
+    static void ParseJSON(VSFileSystem::VSFile &file);
+    static void RecuriveParse(std::map<std::string, std::string> unit_attributes,
+                       const std::string& json_text, bool is_root);
```
Comment (BenjamenMeyer) [inline engine/src/cmd/unit_optimize_factory.h:37]: Probably need `#include <string>` too; would appreciate the add but not going to block per se on just this

**Review by BenjamenMeyer — APPROVED** [review]
Another missing head; would appreciate the add. Otherwise LGTM.
Approving so you don't need to wait further.

**Review by stephengtuggy — APPROVED** [review]
LGTM

**Pull request merged**